File: src/include/asset/AssetType.php
<?php

namespace asset;

enum AssetType: int
{
	case OTHER = 0;
	case PBR_MATERIAL = 1;
	case MODEL_3D = 2;
	case SUBSTANCE_MATERIAL = 3;
	case HDRI = 4;

	public function slug(): string
	{
		return match ($this) {
			AssetType::OTHER => 'other',
			AssetType::PBR_MATERIAL => 'pbr-material',
			AssetType::MODEL_3D => '3d-model',
			AssetType::SUBSTANCE_MATERIAL => 'sbsar',
			AssetType::HDRI => 'hdri',
		};
	}

	public function name(): string
	{
		return match ($this) {
			AssetType::OTHER => 'Other',
			AssetType::PBR_MATERIAL => 'PBR material',
			AssetType::MODEL_3D => '3D model',
			AssetType::SUBSTANCE_MATERIAL => 'Substance material',
			AssetType::HDRI => 'HDRI',
		};
	}

	public static function fromSlug(string $slug): AssetType
	{
		foreach (AssetType::cases() as $c) {
			if ($c->slug() === $slug) {
				return $c;
			}
		}
		return AssetType::OTHER;
	}

	public static function fromTex1Tag(string $tex1Tag): AssetType
	{
		return match ($tex1Tag) {
			"pbr-scanned", "pbr-procedural", "pbr-approximated", "pbr-multiangle", "pbr-stereo" => AssetType::PBR_MATERIAL,
			"sbsar-procedural" => AssetType::SUBSTANCE_MATERIAL,
			"hdri-real" => AssetType::HDRI,
			"3d-modeled", "3d-scanned", "3d-models" => AssetType::MODEL_3D,
			default => AssetType::OTHER,
		};
	}
}

File: internal/model/asset_type.go
package model

// AssetType represents the category of a 3D asset.
type AssetType int

const (
	AssetTypeOther             AssetType = 0
	AssetTypePBRMaterial       AssetType = 1
	AssetTypeModel3D           AssetType = 2
	AssetTypeSubstanceMaterial AssetType = 3
	AssetTypeHDRI              AssetType = 4
)

var allAssetTypes = []AssetType{
	AssetTypeOther,
	AssetTypePBRMaterial,
	AssetTypeModel3D,
	AssetTypeSubstanceMaterial,
	AssetTypeHDRI,
}

func AllAssetTypes() []AssetType { return allAssetTypes }

func (t AssetType) Slug() string {
	switch t {
	case AssetTypeOther:
		return "other"
	case AssetTypePBRMaterial:
		return "pbr-material"
	case AssetTypeModel3D:
		return "3d-model"
	case AssetTypeSubstanceMaterial:
		return "sbsar"
	case AssetTypeHDRI:
		return "hdri"
	default:
		return "other"
	}
}

func (t AssetType) Name() string {
	switch t {
	case AssetTypeOther:
		return "Other"
	case AssetTypePBRMaterial:
		return "PBR material"
	case AssetTypeModel3D:
		return "3D model"
	case AssetTypeSubstanceMaterial:
		return "Substance material"
	case AssetTypeHDRI:
		return "HDRI"
	default:
		return "Other"
	}
}

func AssetTypeFromSlug(slug string) AssetType {
	t, ok := AssetTypeTryFromSlug(slug)
	if !ok {
		return AssetTypeOther
	}
	return t
}

func AssetTypeTryFromSlug(slug string) (AssetType, bool) {
	for _, t := range allAssetTypes {
		if t.Slug() == slug {
			return t, true
		}
	}
	return AssetTypeOther, false
}

func AssetTypeFromTex1Tag(tag string) AssetType {
	switch tag {
	case "pbr-scanned", "pbr-procedural", "pbr-approximated", "pbr-multiangle", "pbr-stereo":
		return AssetTypePBRMaterial
	case "sbsar-procedural":
		return AssetTypeSubstanceMaterial
	case "hdri-real":
		return AssetTypeHDRI
	case "3d-modeled", "3d-scanned", "3d-models":
		return AssetTypeModel3D
	default:
		return AssetTypeOther
	}
}
